lib/mshoplib/src/MShop/Plugin/Provider/Order/PropertyAdd.php

```diff
@@ -121,18 +121,13 @@ public function update( \Aimeos\MW\Observer\Publisher\Iface $order, string $acti
 		if( !is_array( $value ) )
 		{
 			\Aimeos\MW\Common\Base::checkClass( \Aimeos\MShop\Order\Item\Base\Product\Iface::class, $value );
-			return $this->addAttributes( $value, $this->getProductItems( [$value->getProductId()] ), $types );
+			return $this->addAttributes( $value, $this->getProductItems( [$value->getProductId()], [$value->getProductCode()]), $types );
 		}
 
-		$list = [];
+		$ids = map( $value )->getProductId()->toArray();
+		$codes = map( $value )->getProductCode()->toArray();
```

Comment:
You can use `->unique()` already here and pass the Map objects if you change the method signature of getProductsItems() to `iterable` instead of `array`

Reply:
I did do that initially, but my mistake was type hinting as Aimeos\Map, which would have failed for the call we make on the `addAttributes` method outside of the loop.